File: lintcode/find-minimum-in-rotated-sorted-array.py
#!/usr/bin/env python
# encoding: utf-8

# Question: http://www.lintcode.com/en/problem/find-minimum-in-rotated-sorted-array/

class Solution:
    # @param num: a rotated sorted array
    # @return: the minimum number in the array
    def findMin(self, num):
        # write your code here
        if len(num) == 0:
            return 0
        if num[0] < num[-1]:
            return num[0]

        l = 0
        r = len(num) - 1

        while l+1 < r:
            m = l + (r-l)/2
            if num[m] > num[l]:
                l = m
            else:
                r = m
        return min(num[l], num[r])

if __name__ == '__main__':
    s = Solution()
    print s.findMin([1,2,3])

File: leetcode-2/find-minimum-in-rotated-sorted-array.cc
// Question: https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0, r = (int)nums.size()-1;
        while (l < r) {
            int m = l + (r-l)/2;
            if (nums[m] <= nums[l] && nums[m] <= nums[r]) {
                r = m;
            } else if (nums[m] >= nums[l] && nums[m] >= nums[r]) {
                l = m+1;
            } else {
                return nums[l];
            }
        }
        return nums[l];
    }
};
